pkg/source/ttnv2/retry.go

```diff
@@ -0,0 +1,73 @@
+// Copyright © 2021 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ttnv2
+
+import (
+	"context"
+	"time"
+
+	ttnsdk "github.com/TheThingsNetwork/go-app-sdk"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+)
+
+// deviceManagerWithRetry is a ttnsdk.DeviceManager that retries Set() and Get() methods when a resource exhausted or service unavailable error is returned.
+type deviceManagerWithRetry struct {
+	ttnsdk.DeviceManager
+
+	ctx        context.Context
+	maxRetries uint
+	backoff    time.Duration
+}
+
+func (d *deviceManagerWithRetry) shouldRetry(err error, try uint, fields log.Fielder) (bool, time.Duration) {
```

Comment:
```suggestion
func (d *deviceManagerWithRetry) shouldRetry(err error, attempt int, fields log.Fielder) (bool, time.Duration) {
```

Reply:
Also don't we have a context from which we can get a log with fields, like we do everywhere?